src/Microsoft.IdentityModel.Protocols/Configuration/ConfigurationManager.cs

```diff
@@ -157,20 +162,19 @@ public virtual async Task<T> GetConfigurationAsync(CancellationToken cancel)
             //   reach out to the metadata endpoint. Since multiple threads could be calling this method
             //   we need to ensure that only one thread is actually fetching the metadata.
             // else
-            //   if task is running, return the current configuration
+            //   if update task is running, return the current configuration
             //   else kick off task to update current configuration
             if (_currentConfiguration == null)
             {
-                await _configurationNullLock.WaitAsync(cancel).ConfigureAwait(false);
-                if (_currentConfiguration != null)
-                {
-                    _configurationNullLock.Release();
-                    return _currentConfiguration;
-                }
-
-#pragma warning disable CA1031 // Do not catch general exception types
                 try
                 {
+                    await _configurationNullLock.WaitAsync(cancel).ConfigureAwait(false);
```

Comment:
This is a bug. WaitAsync needs to be outside of try.

See example that happened in MSAL: https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/4472

> The semaphore.Wait is called inside a try. When the cancellation token is cancelled, a TaskCancelledException is thrown without the semaphore being acquired, and finally is called trying to release the semaphore which was not acquired causing a SemaphoreFullException.

Reply:
I suppose it's not an issue now because we don't specify max semaphore count, but if we do specify the max count later, this will result in a bug.

Reply:
Nonetheless it shouldn't release something it doesn't hold.

Reply:
Thinking about it more, I think we need to specify a max count of 1 - `SemaphoreSlim(1, 1)`?

Thinking of this scenario:
- Initial semaphore count is 1.
- Thread 1 comes in, takes out a semaphore (semaphore count is 0).
- Thread 2 comes in, waits because semaphore count is 0.
- Thread 1 completes succesfully.
- Thread 1 calls Release in finally (semaphore count is 1).
- Thread 2 moves forward, its cancellation token is expired, so TaskCancelledException is thrown.
  - Semaphore is not taken (semaphore count is still 1).
- Thread 2 goes to catch, then calls release in finally (semaphore count is 2).
- Thread 3 comes in, calls Wait (semaphore count is 1).
- Thread 4 comes in, calls Wait (semaphore count is 0).

Reply:
@pmaytak are you suggesting a race condition where the cancellation token expires while waiting, then the semaphore is not taken?

Reply:
Yes.

Reply:
@pmaytak could you do something like:

```cs
catch (Exception ex) when (ex is not TaskCanceledException)
```